### PR TITLE
DashCanvas

### PR DESCRIPTION
**Pull request by haynesjm42**
#2920 

Toolbox branch: https://github.com/xh/toolbox/tree/reactGridLayout

Beta Checklist

- [x] Support initial width/height (in columns/rows) on DashViewItemSpec
- [x] Better mechanism for adding new views to the layout
- [x] DashView panel renaming support (probably via an option in a menu in the panel `headerItems`)
- [x] Style pass on DashView panel (something more minimal)
- [x] Figure out what we want to call this thing (CardContainer?)

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [ ] Caught up with `develop` branch as of last change.
- [ ] Added CHANGELOG entry, or determined not required.
- [ ] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [ ] Updated doc comments / prop-types, or determined not required.
- [ ] Reviewed and tested on Mobile, or determined not required.
- [ ] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: desktop/cmp/dashGrid/DashGridLayoutContainerModel.js [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+        const {margin, columns: cols, rowHeight, maxRows, ref, containerPadding} = this,
+            containerPosition = ref.current.getBoundingClientRect(),
+            {left: containerLeft, top: containerTop, width: containerWidth} = containerPosition,
+            positionParams = {margin, cols, rowHeight, maxRows,
+                containerPadding: containerPadding ?? margin, containerWidth},
+            left = x - containerLeft,
+            top = y - containerTop;
+
+        this.nextPosition = calcXY(positionParams, top, left);
```
Comment (haynesjm42) [inline desktop/cmp/dashGrid/DashGridLayoutContainerModel.js:293]: See other comment about potentially passing the click position through the context menu so we don't need to keep this temporary value in the model here, and could then calculate this next position at the point of adding the view from the ctx menu.

File: desktop/cmp/dashGrid/DashGridLayoutContainer.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            },
+            item: reactGridLayout({
+                layout: model.layout,
+                cols: model.columns,
+                rowHeight: model.rowHeight,
+                isDraggable: model.isDraggable,
+                isResizable: model.isResizable,
+                compactType: model.compact ? 'vertical' : null,
+                margin: model.margin,
+                maxRows: model.maxRows,
```
[the remark this answers is not in the document]
Comment (ghsolomon) [reply desktop/cmp/dashGrid/DashGridLayoutContainer.js:42]: I found it here under defaultProps (infinity be default):
https://github.com/react-grid-layout/react-grid-layout/blob/master/lib/ReactGridLayout.jsx

File: desktop/cmp/dashGrid/DashGridLayoutContainerModel.js [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        makeObservable(this);
+
+        this.columns = columns;
+        this.rowHeight = rowHeight;
+        this.isDraggable = isDraggable;
+        this.isResizable = isResizable;
+        this.compact = compact;
+        this.initialViews = initialViews;
+        this.restoreState = {initialViews, columns, rowHeight};
+        this.maxRows = Infinity;
```
[the remark this answers is not in the document]
Comment (ghsolomon) [reply desktop/cmp/dashGrid/DashGridLayoutContainerModel.js:52]: Ah yes good point

File: desktop/cmp/dashGrid/DashGridLayoutContainerModel.js [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        const viewModel = this.viewModels.find(it => it.id === id);
+        XH.safeDestroy(viewModel);
+
+        this.viewModels = this.viewModels.filter(it => it.id !== id);
+    }
+
+    @action
+    removeAllViews() {
+        this.layout = [];
+        for (let viewModel of this.viewModels) {
```
[the remark this answers is not in the document]
Comment (ghsolomon) [reply desktop/cmp/dashGrid/DashGridLayoutContainerModel.js:201]: Copy!

File: desktop/cmp/dashGrid/DashGridLayoutContainerModel.js [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    getViewSpec(id) {
+        return this.viewSpecs.find(it => it.id === id);
+    }
+
+    // Get all ViewModels with a given DashViewSpec.id
+    getItemsBySpecId(id) {
+        return this.viewModels.filter(it => it.viewSpec.id === id);
+    }
+
+    setInitialViews() {
```
[the remark this answers is not in the document]
Comment (ghsolomon) [reply desktop/cmp/dashGrid/DashGridLayoutContainerModel.js:232]: Agreed-- that's a better name

File: desktop/cmp/dashGrid/impl/DashGridLayoutView.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    displayName: 'DashGridLayoutView',
+    className: 'xh-dash-tab',
+    model: uses(DashGridLayoutViewModel, {publishMode: ModelPublishMode.LIMITED}),
+
+    render({model, className}) {
+        const {isActive, renderMode, viewSpec, containerModel, id} = model,
+            wasActivated = useRef(false);
+
+        // Respect RenderMode
+        if (!wasActivated.current && isActive) {
```
[the remark this answers is not in the document]
Comment (ghsolomon) [reply desktop/cmp/dashGrid/impl/DashGridLayoutView.js:37]: Makes sense

File: desktop/cmp/dashGrid/DashGridLayoutContainer.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    render({className, model}) {
+        return div({
+            className: classNames(
+                className,
+                model.isDraggable ? `${className}--draggable` : null,
+                model.isResizable ? `${className}--resizable` : null
+            ),
+            ref: model.ref,
+            onContextMenu: (e) => {
+                model.setNextPosition(e.clientX, e.clientY);
```
[the remark this answers is not in the document]
Comment (ghsolomon) [reply desktop/cmp/dashGrid/DashGridLayoutContainer.js:26]: Will look into this

File: package.json [hunk 8]
```diff
@@ -61,9 +61,11 @@
     "react-beautiful-dnd": "~13.1.0",
     "react-dates": "~21.8.0",
     "react-dropzone": "~10.2.2",
+    "react-grid-layout": "^1.3.4",
     "react-onsenui": "~1.11.5",
     "react-popper": "~2.2.5",
     "react-select": "~4.3.1",
+    "react-sizeme": "^3.0.2",
```
Comment (haynesjm42) [inline package.json:68]: Remove

File: desktop/cmp/dash/canvas/DashCanvas.js [hunk 9]
```diff
@@ -6,17 +6,17 @@ import PT from 'prop-types';
...
 
 const reactGridLayout = elemFactory(WidthProvider(ReactGridLayout));
 
-export const [DashReport, dashReport] = hoistCmp.withFactory({
-    displayName: 'DashReport',
-    className: 'xh-dash-report',
-    model: uses(DashReportModel),
+export const [DashCanvas, dashCanvas] = hoistCmp.withFactory({
+    displayName: 'DashCanvas',
+    className: 'xh-dash-Canvas',
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvas.js:25]: Should be lowercase

File: desktop/cmp/dash/canvas/DashCanvas.scss [hunk 10]
```diff
@@ -1,4 +1,4 @@
-.xh-dash-report {
+.xh-dash-Canvas {
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvas.scss:1]: should be lowercase

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+     * This method will clear the persistent state saved for this component, if any.
+     */
+    restoreDefaults() {
+        const {restoreState} = this;
+        this.layoutLocked = restoreState.layoutLocked;
+        this.contentLocked = restoreState.contentLocked;
+        this.renameLocked = restoreState.renameLocked;
+        this.setColumns(restoreState.columns);
+        this.setRowHeight(restoreState.rowHeight);
+        this.setLayout(restoreState.initialState.layout);
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvasModel.js]: Let's just call `setState(restoreState.initialState)` here

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+     * state without any persisted state or user changes applied.
+     *
+     * This method will clear the persistent state saved for this component, if any.
+     */
+    restoreDefaults() {
+        const {restoreState} = this;
+        this.layoutLocked = restoreState.layoutLocked;
+        this.contentLocked = restoreState.contentLocked;
+        this.renameLocked = restoreState.renameLocked;
+        this.setColumns(restoreState.columns);
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvasModel.js:193]: I would just make this method an `@action` and do direct assignments to the bindable properties here for consistency

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    /**
+     * Rename a view in the DashCanvas
+     * @param {string} id - DashCanvasViewModel id to remove from the container
+     */
+    async renameView(id) {
+        const view = this.viewModels.find(it => it.id === id),
+            allowRename = view?.viewSpec?.allowRename;
+        if (!allowRename) return;
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvasModel.js:251]: Should we also be checking `this.renameLocked` here?

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+            ret[id] = {
+                viewSpecId: viewSpec.id,
+                title,
+                viewState
+            };
+        });
+        return ret;
+    }
+
+    set viewState(viewState) {
```
Comment (haynesjm42) [inline desktop/cmp/dash/canvas/DashCanvasModel.js:299]: Let's change this from a property setter to a function (`setViewState(viewState)`) for symmetry with layout - not sure why I set this up this way initially

**Comment by haynesjm42** [comment]
Could we move `renderMode` and `refreshMode` out of `DashViewSpec` and into a new `DashContainerViewSpec` class? (and then update `DashContainerModel` accordingly)

File: desktop/cmp/dash/canvas/DashCanvas.js (hunk 9 above)
Comment (ghsolomon) [reply desktop/cmp/dash/canvas/DashCanvas.js:25]: Thanks for catching these-- casualty of find / replace

File: desktop/cmp/dash/canvas/DashCanvasModel.js (hunk 11 above)
Comment (ghsolomon) [reply desktop/cmp/dash/canvas/DashCanvasModel.js]: I think I'd changed this already-- but I'm running into a bug so let's discuss

File: desktop/cmp/dash/canvas/DashCanvasModel.js [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+     * @param {number} y
+     * @param {number} [w]
+     * @param {number} [h]
+     * @param {Object} viewState
+     * @param {string} title
+     */
+    @action
+    addView(viewSpecId, {x, y, w, h, viewState, title} = {}) {
+        const viewSpec = this.getViewSpec(viewSpecId),
+            id = `${XH.genId()}_${Date.now()}`,
```
[the remark this answers is not in the document]
Comment (ghsolomon) [reply desktop/cmp/dash/canvas/DashCanvasModel.js:212]: Sounds good

File: desktop/cmp/dash/canvas/DashCanvasModel.js (hunk 14 above)
Comment (ghsolomon) [reply desktop/cmp/dash/canvas/DashCanvasModel.js:299]: I might have changed this, but actually caught the discrepancy too and refactored

File: desktop/cmp/dash/canvas/impl/DashCanvasView.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Implementation component to show an item within a DashCanvas.  This component
+ * is used by DashCanvas's internal implementation to:
+ *
+ *   - Mount/unmount its contents according to `DashCanvasViewSpec.renderMode`.
+ *   - Track and trigger refreshes according to `DashCanvasViewSpec.refreshMode`.
+ *   - Stretch its contents using a flex layout.
+ *
+ * @private
+ * @Beta
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/impl/DashCanvasView.js:25]: Don't think @Beta is real, and in any case not needed for this if its private anyway.

File: desktop/cmp/dash/canvas/impl/DashCanvasContextMenu.js [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+import {hoistCmp} from '@xh/hoist/core';
+import {contextMenu} from '@xh/hoist/desktop/cmp/contextmenu';
+import {Icon} from '@xh/hoist/icon';
+import {isEmpty} from 'lodash';
+
+/**
+ * Context menu to add items to a DashCanvas
+ *
+ * -------- !! NOTE: THIS COMPONENT IS CURRENTLY IN BETA !! --------
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/impl/DashCanvasContextMenu.js:16]: Beta warning not needed here --- actually, just want on the main public entry points. - Component and model

File: desktop/cmp/dash/canvas/impl/DashCanvasContextMenu.js [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    const addMenuItems = createAddMenuItems({dashCanvasModel, clickPosition}),
+        {extraMenuItems, contentLocked} = dashCanvasModel;
+    return [
+        {
+            text: 'Add',
+            icon: Icon.add(),
+            disabled: contentLocked || isEmpty(addMenuItems),
+            items: addMenuItems
+        },
+        ...(extraMenuItems ? ['-', ...extraMenuItems] : [])
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/impl/DashCanvasContextMenu.js:48]: this does not need to be so nested -- should be able to be:

'-',
...(extramenuItems ?? [])


extraneous '-' will be dropped if needed.

File: desktop/cmp/dash/canvas/impl/DashCanvasView.js [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+                            },
+                            {
+                                text: 'Remove',
+                                icon: Icon.cross(),
+                                intent: 'danger',
+                                hidden: !viewSpec.allowRemove,
+                                disabled: contentLocked,
+                                actionFn: () => containerModel.removeView(id)
+                            },
+                            ...(extraMenuItems ? ['-', ...extraMenuItems] : [])
```
Comment (lbwexler) [inline desktop/cmp/dash/canvas/impl/DashCanvasView.js:72]: extra level of nesting

**Pull request merged**